ibm/service/database/resource_ibm_database.go

```diff
@@ -2836,73 +2799,264 @@ func checkV5Groups(_ context.Context, diff *schema.ResourceDiff, meta interface{
 	return nil
 }
 
-// Updates and creates users. Because we cannot get users, we first attempt to update the users, then create them
-func userUpdateCreate(userData map[string]interface{}, instanceID string, meta interface{}, d *schema.ResourceData) (err error) {
-	cloudDatabasesClient, _ := meta.(conns.ClientSession).CloudDatabasesV5()
+func validateUsersDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) (err error) {
+	oldUsers, newUsers := diff.GetChange("users")
+	userChanges := expandUserChanges(oldUsers.(*schema.Set).List(), newUsers.(*schema.Set).List())
+
+	for _, change := range userChanges {
+		if change.isDelete() {
+			continue
+		}
+
+		if change.isCreate() || change.isUpdate() {
+			err = change.New.Validate()
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func expandUsers(_users []interface{}) []*DatabaseUser {
+	if len(_users) == 0 {
+		return nil
+	}
+
+	users := make([]*DatabaseUser, 0, len(_users))
+
+	for _, userRaw := range _users {
+		if tfUser, ok := userRaw.(map[string]interface{}); ok {
+
+			user := DatabaseUser{
+				Username: tfUser["name"].(string),
+				Password: tfUser["password"].(string),
+				Role:     tfUser["role"].(string),
+				Type:     tfUser["type"].(string),
+			}
+
+			users = append(users, &user)
+		}
+	}
+
+	return users
+}
+
+func expandUserChanges(_oldUsers []interface{}, _newUsers []interface{}) (userChanges []*userChange) {
+	oldUsers := expandUsers(_oldUsers)
+	newUsers := expandUsers(_newUsers)
+
+	userChangeMap := make(map[string]*userChange)
+
+	for _, user := range oldUsers {
+		userChangeMap[user.ID()] = &userChange{Old: user}
+	}
+
+	for _, user := range newUsers {
+		if _, ok := userChangeMap[user.ID()]; !ok {
+			userChangeMap[user.ID()] = &userChange{}
+		}
+		userChangeMap[user.ID()].New = user
+	}
+
+	userChanges = make([]*userChange, 0, len(userChangeMap))
+
+	for _, user := range userChangeMap {
+		userChanges = append(userChanges, user)
+	}
+
+	return userChanges
+}
+
+func (c *userChange) isDelete() bool {
+	return c.Old != nil && c.New == nil
+}
+
+func (c *userChange) isCreate() bool {
+	return c.Old == nil && c.New != nil
+}
+
+func (c *userChange) isUpdate() bool {
+	return c.New != nil && c.Old != nil && c.Old.Password != c.New.Password
+}
```

Comment:
This considers password change as an update. What about changes to name/type/role ?

I understand some of them will cause delete and recreate. How are they handled?

Reply:
Users are grouped by name and type in the `expandUserChanges` method. Any change to name or type will delete the old user and create a new one. 

Ops Manager role updates are not currently supported in the provider but updating roles will be added with Redis RBAC support.